### PR TITLE
Add fix command for extracting using directives into `project.scala` file

### LINKED ISSUE #2104: Provide a way to extract all directives declared across different scala files into a single file
**Is your feature request related to a problem? Please describe.**
I found that allowing directives all over the place got confusing really quick (but I understand its good for prototyping). It would be extremely useful to have a command in order to centralize all directives into a single file (like project.scala)

**Describe the solution you'd like**
Run a command like `scala-cli --centralize-directives` in order for the process to happen automatically

**Describe alternatives you've considered**
Maniually going through each file and removing the directives from there and placing them into `project.scala`

**Additional context**
Currently, Scala CLI produces a warning with suggestions to place the directives in `project.scala` - we need to make it actionable (so you can click and apply the changes or run a separate command in order to do this).

https://twitter.com/cal_fern/status/1655912142753677314

### PR DESCRIPTION
**Pull request by MaciejG604**
Fixes #2104

Introduces a new subcommand `fix` that extracts all `using directives` that exist througout the project sources (not only files on disk but everything that's linked to the project build). Those extracted directives are then written to `project.scala` file.

This process can be split into two parts (their order is not exactly 1. -> 2. in code):
1. Extraction of directives and writing them into `project.scala`
2. Modification of directives in their original sources.

Rules regarding the 1. part:
- if a source contains any `target using directive` then it's directives are not extracted, since it's a correct way to create projects with cross-configuration (the warning about using directives in multiple files has been modified accordingly)
- directives from sources belonging to the Test scope are extracted and written to `project.scala` using their `test.` equivalents (using dep -> using test.dep) if possible, if not possible they are not written to `project.scala`

Rules regarding the 2. part: 
- if a source is writable (is on local disk) and it is a member of original inputs then the extracted directives are removed from its contents, otherwise no modification takes place
- if a source belonging to the Test scope contains directives that do not have `test.` equivalents they are kept in this source and are not moved to `project.scala`, all the others extracted `using directives` are removed from the source if rule above applies

### CODE REVIEW

**Comment by mpkocher** [comment]
Where are the requirements of this defined? I'm not really seeing anything in the original issue. 

You're adding a new subcommand called "fix"?

**Comment by MaciejG604** [comment]
> Where are the requirements of this defined? I'm not really seeing anything in the original issue.

The requirements for this were discussed internally by the Scala CLI team. I can list them in the description since they are quite simple.

> You're adding a new subcommand called "fix"?

Yes, the proposal right now is to add this functionaility in a `fix` subcommand. However, all suggestions are highly appreciated.

File: modules/directives/src/main/scala/scala/build/preprocessing/directives/DirectiveHandler.scala [hunk 1]
```diff
@@ -36,8 +36,8 @@ trait DirectiveHandler[+T] { self =>
 
   final def isRestricted: Boolean   = scalaSpecificationLevel == SpecificationLevel.RESTRICTED
   final def isExperimental: Boolean = scalaSpecificationLevel == SpecificationLevel.EXPERIMENTAL
-
-  def keys: Seq[String]
+  // Sequence of names and aliases, those beloning to the same values are grouped in inner sequences
```
Comment (lwronski) [inline modules/directives/src/main/scala/scala/build/preprocessing/directives/DirectiveHandler.scala]:
```
def keys: Seq[Keys]
```

What do you think about creating class to expose names and aliases? If not, at least this comment should be the ScalaDoc.

File: modules/build/src/main/scala/scala/build/preprocessing/UsingDirectivesOps.scala [hunk 2]
```diff
@@ -12,8 +12,9 @@ object UsingDirectivesOps {
     def containsTargetDirectives: Boolean = ud.keySet.exists(_.startsWith("target."))
 
     def getPosition(path: Either[String, os.Path]): Position.File =
-      val line   = ud.getAst().getPosition().getLine()
-      val column = ud.getAst().getPosition().getColumn()
+      // TODO Use the position from the AST instead when fixed in using-directives
+      val line   = -1
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/preprocessing/UsingDirectivesOps.scala]: It's worth dedicating some time to fix this issue. Setting the line to `-1` can be confusing.

File: modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    val scopedSources = crossSources.scopedSources(sharedOptions).orExit(logger)
+
+    val mainSources = scopedSources.sources(Scope.Main, sharedOptions)
+    val testSources = scopedSources.sources(Scope.Test, sharedOptions)
+
+    def isExtractedFromWritableInput(position: Option[Position.File]): Boolean = {
+      // Only initial inputs are used, new inputs discovered during processing of
+      // CrossSources.forInput may be shared between projects
+      val writableInputs: Seq[OnDisk] = inputs.flattened()
+        .collect { case onDisk: OnDisk => onDisk }
```
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala]: The initialization of this variable should be before the body of this function.

File: modules/cli/src/main/scala/scala/cli/commands/fix/FixOptions.scala [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  private val helpHeader  = "Perform fixes on a Scala CLI project."
+  val helpMessage: String = HelpMessages.shortHelpMessage(cmdName, helpHeader)
+  val detailedHelpMessage: String =
+    s"""$helpHeader
+       |
+       |${HelpMessages.commandConfigurations(cmdName)}
+       |
+       |${HelpMessages.acceptedInputs}
+       |
+       |To pass arguments to the actual application, just add them after `--`, like:
```
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/fix/FixOptions.scala]: It appears that this message was copied from `RunOptions`. However, in the `fix` option message, it's not useful.

File: modules/cli/src/main/scala/scala/cli/commands/fix/FixOptions.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  @Tag(tags.experimental)
+    migrateDirectives: Option[Boolean] = None,
+//  @HelpMessage(s"Perform fixes on main scope")
+//  @Tag(tags.inShortHelp)
+//  @Tag(tags.experimental)
+//    mainScope: Option[Boolean] = None,
+//  @HelpMessage(s"Perform fixes on test scope")
+//  @Tag(tags.inShortHelp)
+//  @Tag(tags.experimental)
+//    testScope: Option[Boolean] = None
```
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/fix/FixOptions.scala]: I think this option isn't necessary and can be removed.

File: modules/integration/src/test/scala/scala/cli/integration/FixTests.scala [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+
+      val projectFileContents = os.read(root / projectFileName)
+      val mainFileContents    = os.read(root / mainSubPath)
+      val testFileContents    = os.read(root / testSubPath)
+
+      assertNoDiff(
+        projectFileContents,
+        """// Main
+          |//> using objectWrapper
+          |//> using dependency "com.lihaoyi::os-lib:0.9.1", "com.lihaoyi::pprint:0.6.6"
```
Comment (lwronski) [inline modules/integration/src/test/scala/scala/cli/integration/FixTests.scala]: I think the `fix` command shouldn't generate directives that are separated by commas. It should enough to specify values without using commas.

File: modules/build/src/main/scala/scala/build/preprocessing/DirectivesPreprocessor.scala [hunk 7]
```diff
@@ -170,7 +170,7 @@ object DirectivesPreprocessor {
 
     val handlersMap = handlers
       .flatMap { handler =>
-        handler.keys.map(k => k -> handleValues(handler))
+        handler.keys.flatten.map(k => k -> handleValues(handler))
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/preprocessing/DirectivesPreprocessor.scala]:
```suggestion
        handler.keys.flatMap(_.nameAliases).map(k => k -> handleValues(handler))
```

File: modules/directives/src/main/scala/scala/build/preprocessing/directives/DirectiveHandler.scala [hunk 8]
```diff
@@ -80,6 +80,12 @@ trait DirectiveHandler[+T] { self =>
 
 }
 
+/** Using directive key with all its aliases */
+case class Key(nameAliases: Seq[String])
+object Key {
+  given Conversion[Key, IterableOnce[String]] = _.nameAliases
+}
+
```
Comment (lwronski) [inline modules/directives/src/main/scala/scala/build/preprocessing/directives/DirectiveHandler.scala]:
```suggestion
```

File: modules/build/src/main/scala/scala/build/preprocessing/UsingDirectivesOps.scala [hunk 9]
```diff
@@ -12,9 +13,33 @@ object UsingDirectivesOps {
...
+      extension (pos: Positioned) {
+        def getLine   = pos.getPosition.getLine
+        def getColumn = pos.getPosition.getColumn
+      }
+
+      @tailrec
+      def getEndPostion(ast: UsingTree): (Int, Int) = ast match {
+        case uds: UsingDefs => uds.getUsingDefs.asScala.toSeq match {
+            case Nil => (uds.getLine, uds.getColumn)
+            case seq => getEndPostion(seq.last)
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/preprocessing/UsingDirectivesOps.scala]:
```suggestion
            case _ :+ lastUsingDef => getEndPostion(lastUsingDef)
```

File: modules/build/src/main/scala/scala/build/preprocessing/UsingDirectivesOps.scala [hunk 10]
```diff
@@ -12,9 +13,33 @@ object UsingDirectivesOps {
...
+          }
+        case ud: UsingDef => getEndPostion(ud.getValue)
+        case uvs: UsingValues => uvs.getValues.asScala.toSeq match {
+            case Nil => (uvs.getLine, uvs.getColumn)
+            case seq => getEndPostion(seq.last)
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/preprocessing/UsingDirectivesOps.scala]:
```suggestion
            case _ :+ lastUsingDef => getEndPostion(lastUsingDef)
```

File: modules/integration/src/test/scala/scala/cli/integration/FixTests.scala [hunk 11]
```diff
@@ -0,0 +1,247 @@
+package scala.cli.integration
+
+import com.eed3si9n.expecty.Expecty.expect
+
+class FixTests extends ScalaCliSuite {
```
Comment (lwronski) [inline modules/integration/src/test/scala/scala/cli/integration/FixTests.scala:5]: Add integration test with `//> using file` directive.

File: modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    {
+      val originalMainDirectives = getExtractedDirectives(mainSources, logger)
+        .filterNot(hasTargetDirectives)
+
+      val transformedMainDirectives = unifyCorrespondingNameAliases(originalMainDirectives)
+
+      val directivesFromWritableMainInputs = transformedMainDirectives
+        .filter(d => isExtractedFromWritableInput(d.positions))
+
+      directivesFromWritableMainInputs.foreach(d => removeDirectivesFrom(d.positions))
```
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala]: Removing directives should be move at the end of execution `Fix` command. The same for `test` scope.

File: modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala [hunk 13]
```diff
@@ -217,11 +226,21 @@ object Fix extends ScalaCommand[FixOptions] {
...
+  )(
+    using loggingUtilities: LoggingUtilities
   ): Unit = {
     position match {
       case Some(Position.File(Right(path), _, _, offset)) =>
-        val keepLines   = toKeep.mkString("", newLine, newLine * 2)
-        val newContents = keepLines + os.read(path).drop(offset).stripLeading()
+        val keepLines    = toKeep.mkString("", newLine, newLine * 2)
+        val newContents  = keepLines + os.read(path).drop(offset).stripLeading()
+        val relativePath = loggingUtilities.relativePath(path)
```
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala:264]: What path will be printed if users pass:
```
$ echo $pwd
/tmp/scala-demo
$ scala-cli fix Hello.scala /User/scala-cli/project/scala-cli/Hello.scala
```

Is it proper to show a relative path for the second path?

Comment (MaciejG604) [reply modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala:264]: Good point, the second path gets turned into an ugly `../../User/scala-cli/project/scala-cli/Hello.scala` so a fix is needed.

File: modules/build/src/main/scala/scala/build/preprocessing/SheBang.scala [hunk 14]
```diff
@@ -7,6 +7,19 @@ object SheBang {
 
   def isShebangScript(content: String): Boolean = sheBangRegex.unanchored.matches(content)
 
+  /** Returns the shebang section and the content without the shebang section */
+  def partitionOnShebangSection(content: String): (String, String) =
+    if (content.startsWith("#!")) {
+      val regexMatch = sheBangRegex.findFirstMatchIn(content)
+      regexMatch match {
+        case Some(firstMatch) =>
+          (firstMatch.toString(), content.replace(firstMatch.toString(), ""))
```
Comment (lwronski) [inline modules/build/src/main/scala/scala/build/preprocessing/SheBang.scala]:
```suggestion
          (firstMatch.toString(), content.replaceFirst(firstMatch.toString(), ""))
```

File: modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+        buildOptions.archiveCache,
+        buildOptions.internal.javaClassNameVersionOpt,
+        () => buildOptions.javaHome().value.javaCommand
+      ),
+      logger = logger,
+      suppressWarningOptions = SuppressWarningOptions(
+        Some(true),
+        Some(true),
+        Some(true)
+      ),
```
Comment (lwronski) [inline modules/cli/src/main/scala/scala/cli/commands/fix/Fix.scala]:
```suggestion
      suppressWarningOptions = SuppressWarningOptions.suppressAll()
```

If a new field is added to `SuppressWarningOptions` with a default value, this code will have to be updated as well. Perhaps we should create a variable in `SuppressWarningOptions` object that suppresses all warnings

**Review by lwronski — APPROVED** [review]

**Pull request merged**